go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -558,3 +570,298 @@ func recalculatePKColsInfoByColumnNames(uniqueKeyColumnNames []string, colInfos
+		if err != nil {
+			return err
+		}
+		// Use a new DB client to avoid interfering with open transactions
+		// in the shared client as DDL includes an implied commit.
+		// We're also NOT using a DBA connection here because we want to
+		// be sure that the commit fails if the instance is somehow in
+		// READ-ONLY mode.
+		dbClient := vr.vre.dbClientFactoryFiltered()
+		if err := dbClient.Connect(); err != nil {
+			log.Errorf("unable to connect to the database when saving secondary keys for deferred creation: %v", err)
+			return err
+		}
+		defer dbClient.Close()
+		if _, err := dbClient.ExecuteFetch(insert, 1); err != nil {
+			return err
+		}
+		if _, err := dbClient.ExecuteFetch(sqlparser.String(alterDrop), 1); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (vr *vreplicator) getTableSecondaryKeys(ctx context.Context, tableName string) ([]*sqlparser.IndexDefinition, error) {
+	req := &tabletmanagerdatapb.GetSchemaRequest{Tables: []string{tableName}}
+	schema, err := vr.mysqld.GetSchema(ctx, vr.dbClient.DBName(), req)
+	if err != nil {
+		return nil, err
+	}
+	if len(schema.TableDefinitions) != 1 {
+		return nil, fmt.Errorf("unexpected number of table definitions returned from GetSchema call for table %q: %d",
+			tableName, len(schema.TableDefinitions))
+	}
+	ddl := schema.TableDefinitions[0].Schema
+	var secondaryKeys []*sqlparser.IndexDefinition
+	parsedDDL, err := sqlparser.ParseStrictDDL(ddl)
+	var foundCreateTable bool
+	if err != nil {
+		return secondaryKeys, err
+	}
+
+	err = sqlparser.Walk(func(node sqlparser.SQLNode) (kontinue bool, err error) {
+		switch node := node.(type) {
+		case sqlparser.DDLStatement:
+			foundCreateTable = true
+			if node.GetTableSpec() != nil {
+				for _, index := range node.GetTableSpec().Indexes {
+					if !index.Info.Primary {
+						secondaryKeys = append(secondaryKeys, index)
+					}
+				}
+			}
+		}
+		return true, nil
+	}, parsedDDL)
+
+	if !foundCreateTable {
+		return nil, fmt.Errorf("could not find expected CREATE TABLE statement in DDL: %s", ddl)
+	}
+
+	return secondaryKeys, err
+}
+
+func (vr *vreplicator) execPostCopyActions(ctx context.Context, tableName string) error {
+	// Use a new DB client to avoid interfering with open transactions
+	// in the shared client as DDL includes an implied commit.
+	// We're also NOT using a DBA connection here because we want to be
+	// sure that the work fails if the instance is somehow in READ-ONLY
+	// mode.
+	dbClient := vr.vre.dbClientFactoryFiltered()
+	if err := dbClient.Connect(); err != nil {
+		return fmt.Errorf("unable to connect to the database when executing post copy actions: %v", err)
+	}
+	defer dbClient.Close()
+
+	query, err := sqlparser.ParseAndBind(sqlGetPostCopyActions, sqltypes.Uint32BindVariable(vr.id),
+		sqltypes.StringBindVariable(tableName))
+	if err != nil {
+		return err
+	}
+	qr, err := dbClient.ExecuteFetch(query, -1)
+	if err != nil {
+		return err
+	}
+	if qr == nil || len(qr.Rows) == 0 {
```

Comment:
In what situation would `qr == nil`?

Reply:
It should never happen, just an extra check to prevent some unexpected situation from causing a crash. I'll add a comment to note that it should never happen.